西暦1400年前後に活躍したインドの数学者、マーダヴァは以下の公式を発見したとされています。

$$
\begin{align}
\frac{\pi}{4} &= \sum_{k = 0}^{\infty} (-1)^{k} \frac{1}{2k + 1} \\
&= 1 - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} - \cdots
\end{align}
$$

[![alt設定](http://img.youtube.com/vi/O5XECJniaNE/0.jpg)](https://www.youtube.com/watch?v=O5XECJniaNE)

後にグレゴリー、ライプニッツが独立に同じ公式を発見しており、右辺の級数はマーダヴァ・グレゴリー・ライプニッツ級数と呼ばれます。この公式を用いて円周率を求めてみましょう。

右辺の級数は足し算と引き算を交互に繰り返しており、$k$ が奇数の項まで計算すると、

$$
\begin{align}
\frac{\pi}{4} &= \sum_{k = 0}^{2n - 1} (-1)^{k} \frac{1}{2k + 1} + \sum_{k=2n}^{\infty} (-1)^{k} \frac{1}{2k + 1} \\
&= \sum_{k = 0}^{2n - 1} (-1)^{k} \frac{1}{2k + 1} + \sum_{m=n}^{\infty} \left( \frac{1}{4m + 1} - \frac{1}{4m + 3} \right) \\
&= \sum_{k = 0}^{2n - 1} (-1)^{k} \frac{1}{2k + 1} + \sum_{m=n}^{\infty} \frac{2}{(4m + 1)(4m + 3)} \\
&> \sum_{k = 0}^{2n - 1} (-1)^{k} \frac{1}{2k + 1}
\end{align}
$$

であり、$k$ が偶数の項まで計算すると

$$
\begin{align}
\frac{\pi}{4} &= \sum_{k = 0}^{2n} (-1)^{k} \frac{1}{2k + 1} + \sum_{k=2n+1}^{\infty} (-1)^{k} \frac{1}{2k + 1} \\
&= \sum_{k = 0}^{2n} (-1)^{k} \frac{1}{2k + 1} + \sum_{m=n}^{\infty} \left(- \frac{1}{4m + 3} + \frac{1}{4m + 5} \right) \\
&= \sum_{k = 0}^{2n} (-1)^{k} \frac{1}{2k + 1} - \sum_{m=n}^{\infty} \frac{2}{(4m + 3)(4m + 5)} \\
&< \sum_{k = 0}^{2n} (-1)^{k} \frac{1}{2k + 1}
\end{align}
$$

なので、

$$\sum_{k = 0}^{2n - 1} (-1)^{k} \frac{1}{2k + 1} < \frac{\pi}{4} < \sum_{k = 0}^{2n} (-1)^{k} \frac{1}{2k + 1}$$

となります。これによって $\pi$ が上下から評価され、$\pi$ の値を求めることができます。

In [93]:
def madhava_series(num_of_terms: int) -> float:
    val = 0
    for i in range(0, num_of_terms):
        val = val + (-1) ** i * (1 / (2 * i + 1))
        
    return 4 * val

In [92]:
terms = [2, 10, 10**2, 10**3, 10**4, 10**5]

for t in terms:
    print(f"{t}項目 + 1まで計算")
    pi_low = madhava_series(t)
    pi_high = madhava_series(t+1)
    print(f"{pi_low} <")
    print(f"{pi_high}\n")

2項目 + 1まで計算
2.666666666666667 <
3.466666666666667

10項目 + 1まで計算
3.0418396189294032 <
3.232315809405594

100項目 + 1まで計算
3.1315929035585537 <
3.1514934010709914

1000項目 + 1まで計算
3.140592653839794 <
3.1425916543395442

10000項目 + 1まで計算
3.1414926535900345 <
3.1416926435905346

100000項目 + 1まで計算
3.1415826535897198 <
3.1416026534897203



円周率は　$\pi = 3.1415926535897932 \cdots$ なので、100000 項目まで計算しても 4 桁までしか求まりません (とはいえ、小数点第5桁目以降の6桁目から13桁目まで一致しているのは不思議です)。つまり収束が非常に遅いです。

マーダヴァは以下のように補正項を入れることで、収束を速めていたようです。

$$F_2(n) = \frac{n}{4n^2 + 1}$$

$$\frac{\pi}{4} \fallingdotseq \sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n} F_2(n)$$

他にも、

$$F_3(n) = \frac{n^2 + 1}{4n^3 + 5n}$$

として、

$$\frac{\pi}{4} \fallingdotseq \sum_{k=0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n} F_3(n)$$
    
という近似も考えていたようです。どのように補正項を求めたのかは動画を見てください。

これが本当に $\pi$ に収束するのか、収束が速くなるのかを確認してみましょう。

$$C_2(n) = \sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n} F_2(n)$$

$$C_3(n) = \sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n} F_3(n)$$

とおきます。このとき、

$$
\begin{align}
\lim_{n \to \infty} C_2(n) &= \lim_{n \to \infty}\sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + \lim_{n\to \infty} F_2(n)
&= \frac{\pi}{4} + 0
\end{align}
$$

なので $C_2(n)$ は $\frac{\pi}{4}$ に収束します。同様に $C_3(n)$ も収束します。

ここで、$C_2(n+1) -C_2(n)$ を計算すると、

$$\begin{align}
C_2(n+1) - C_2(n) &= \sum_{k = 0}^{n} (-1)^{k} \frac{1}{2k +1} + (-1)^{n+1} F_2(n+1) - \left(\sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n}F_2(n)\right) \\
&= (-1)^{n}\left(\frac{1}{2n + 1} -  F_2(n+1) - F_2(n) \right) \\
&= (-1)^{n}\left(\frac{1}{2n + 1} - \left(\frac{n + 1}{4n^2 + 4n + 5} + \frac{n}{4m^2 + 1}\right)\right) \\
&= (-1)^n \frac{4}{(2n + 1)(4n^2 + 4n + 5)(4n^2 + 1)}
\end{align}$$

となります。よって $n$ が偶数なら $C_2(n) < C_2(n+1)$、$n$ が奇数なら $C_2(n) > C_2(n+1)$ となります。

また、$C_2(n + 2) - C_2(n)$ を計算すると、

$$\begin{align}
C_2(n+2) - C_2(n) &= \sum_{k = 0}^{n+1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n+2} F_2(n+2) - \left(\sum_{k = 0}^{n-1} (-1)^{k} \frac{1}{2k +1} + (-1)^{n} F_2(n)\right) \\
&= \frac{(-1)^{n}}{2n + 1} + \frac{(-1)^{n+1}}{2n + 3} + (-1)^{n} \left(F_2(n+2) - F_2(n)\right) \\
&= (-1)^{n}\left(\frac{2}{(2n+1)(2n+3)} + \frac{n + 2}{(4(n+2)^2 + 1)} - \frac{n}{4n^2 + 1}\right) \\
&= (-1)^{n}\left(\frac{2}{(2n+1)(2n+3)} - \frac{8n^2 + 16n -2}{(4n^2 + 16n + 17)( 4n^2 + 1)}\right) \\
&= (-1)^{n}\left(\frac{6}{(2n+1)(2n+3)(4n^2 + 16n + 17)( 4n^2 + 1)}\right) \\
\end{align}$$

なので、$n$ が偶数なら $C_2(n + 2) > C_2(n)$、$n$ が奇数なら $C_2(n + 2) < C_2(n)$ となります。したがって

$$C_2(2) < C_2(4) < \cdots < C_2(2m) < \cdots < \frac{\pi}{4} < \cdots < C_2(2m+1) < \cdots C_2(3) < C_2(1)$$

が成り立ちます。

$C_3(n)$ も同様のことが成り立ちます。

それではこの補正項で収束が速くなるのか確かめてみましょう。この記事では、面倒なので片側のみの計算をします。

In [101]:
import numpy as np
import mpmath
mpmath.mp.dps = 50

PI_50 = "3.14159265358979323846264338327950288419716939937510"

def f1(n: int) -> float:
    return 1/(4*n)

def f2(n: int) -> float:
    return n / (4 * n ** 2 + 1)

def f3(n: int) -> float:
    return (n ** 2 + 1) / (4 * n ** 3 + 5 * n)

def madhava(num_of_terms: int, correction_term) -> float:
    val = mpmath.mpf("0")
    for i in range(0, num_of_terms):
        _i = mpmath.mpf(f"{i}")
        val = val + (-1) ** i * (1 / (2 * _i + 1))
        
    n = mpmath.mpf(f"{num_of_terms}")
        
    val = val + (-1) ** (num_of_terms) * correction_term(n)
    return 4 * val

def print_pi(pi):
    pi_str = mpmath.nstr(pi, 50, strip_zeros=False)
    
    def match_nod(pi_str: str) -> int:
        for i in range(0, min(len(pi_str), len(PI_50))):
            if pi_str[i] != PI_50[i]:
                break
        return i 
    
    nod = match_nod(pi_str)
    
    print(f"({max(0, nod -2)}桁まで一致)")
    print(f"\033[31m{pi_str[0:nod]}\033[0m{pi_str[nod:]} \n")

In [104]:
terms = [1, 10, 10**2, 10**3, 10**4, 10**5]

for t in terms:
    print(f"=== {t:,}項目まで計算 ===")
    
    print("・ 補正なし")
    pi = madhava(t, lambda x: 0)
    print_pi(pi)
    
    print("・ F1による補正")
    pi = madhava(t, f1)
    print_pi(pi)
    
    print("・ F2による補正")
    pi = madhava(t, f2)
    print_pi(pi)
    
    print("・ F3による補正")
    pi = madhava(t, f3)
    print_pi(pi)

=== 1項目まで計算 ===
・ 補正なし
(0桁まで一致)
4.0000000000000000000000000000000000000000000000000 

・ F1による補正
(0桁まで一致)
3.0000000000000000000000000000000000000000000000000 

・ F2による補正
(0桁まで一致)
3.2000000000000000000000000000000000000000000000000 

・ F3による補正
(1桁まで一致)
3.1111111111111111111111111111111111111111111111111 

=== 10項目まで計算 ===
・ 補正なし
(0桁まで一致)
3.0418396189294022111359572659882257405477219718706 

・ F1による補正
(3桁まで一致)
3.1418396189294022111359572659882257405477219718706 

・ F2による補正
(5桁まで一致)
3.1415902423707987198641368171104202542634326950626 

・ F3による補正
(6桁まで一致)
3.1415927053491552975557103524079788269674750582903 

=== 100項目まで計算 ===
・ 補正なし
(1桁まで一致)
3.1315929035585527643074142382769205164030543844066 

・ F1による補正
(6桁まで一致)
3.1415929035585527643074142382769205164030543844066 

・ F2による補正
(10桁まで一致)
3.1415926535648026080613203906231118616194239751668 

・ F3による補正
(14桁まで一致)
3.1415926535897988585456344607491114925310383864063 

=== 1,000項目まで計算 ===
・ 補正なし
(2桁まで一致)
3.140592653839792925963596502869395970451389

円周率 50桁までの値は

$$\pi = 3.14159265358979323846264338327950288419716939937510$$

なので、補正項を入れることで劇的に収束が速くなることがわかります。

ちなみに、マーダヴァは円周率の近似値を

$$\pi \fallingdotseq \frac{2827433388233}{9 \times 10^{11}}$$

としたとされています。これをどのように求めたのでしょうか。

In [65]:
from fractions import Fraction

def f2_frac(n: int) -> Fraction:
    return Fraction(n, 4 * n ** 2 + 1)

def f3_frac(n: int) -> Fraction:
    return Fraction(n ** 2 + 1, 4 * n ** 3 + 5 * n)

def madhava_fraction(num_of_terms: int, correction_term) -> Fraction:
    val = Fraction(0)
    for i in range(1, num_of_terms + 1):
        val = val + (-1) ** (i - 1) * Fraction(1, 2 * i - 1)
        
    val = val + (-1) ** (num_of_terms) * correction_term(num_of_terms)
        
    return 4 * val

In [99]:
print(f"2827433388233 / (9 × 10^11) = {2827433388233 / (9 * 10 ** 11)}\n")

for i in range(30, 45):
    print(f"=== {i}桁目 ===")
    # print("・　補正なし")
    # pi = madhava_fraction(i, lambda x: 0)
    # print(f"{pi} = {float(pi)}")
    # print("・　F2補正")
    # pi = madhava_fraction(i, f2_frac)
    # print(f"{pi} = {float(pi)}")
    print("・　F3補正")
    pi = madhava_fraction(i, f3_frac)
    print(f"{pi} = {float(pi)}\n")

2827433388233 / (9 × 10^11) = 3.1415926535922223

=== 30桁目 ===
・　F3補正
97992494529747406341952274/31191979780376714537313075 = 3.141592653615266

=== 31桁目 ===
・　F3補正
74458122323009486535693376972/23700756442248765144016827075 = 3.1415926535695324

=== 32桁目 ===
・　F3補正
634664088646835919864811387499/202019853821069754961003913400 = 3.1415926536060255

=== 33桁目 ===
・　F3補正
397312362057854977366893772/126468452746575939673691325 = 3.1415926535767

=== 34桁目 ===
・　F3補正
461511789912508589897259063338/146903765319031092352875088425 = 3.1415926536004224

=== 35桁目 ===
・　F3補正
32601934608006633933297110972/10377518094474652667831098275 = 3.1415926535811125

=== 36桁目 ===
・　F3補正
110194239875299492609058627352929/35075915952736303314033627480525 = 3.141592653596923

=== 37桁目 ===
・　F3補正
1550236945627968928958135268428/493455745721670869517142957425 = 3.1415926535839054

=== 38桁目 ===
・　F3補正
1550236945633285619163512942782/493455745721670869517142957425 = 3.14159265359468

=== 39桁目 ===
・　F3補正
122712105905

$F_3$ による補正で42桁目まで計算した値は 

$$3.141592653592221 \cdots$$

であり、
 
$$\frac{2827433388233}{9 \times 10^{11}} \fallingdotseq 3.1415926535922223$$

にかなり近いです。よってマーダヴァは42桁目あたりまで計算し、分母を綺麗にするために適当に整理したのではないかと思われますが、分母の数字があまりにも異なるので、どのように整理したのかはわかりません。